cni/pkg/ebpf/server/redirectionServer.go

```diff
@@ -0,0 +1,542 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang -target bpf -cflags "-D__TARGET_ARCH_x86"  ambient_redirect ../app/ambient_redirect.bpf.c
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/cilium/ebpf"
+	"github.com/florianl/go-tc"
+	"github.com/florianl/go-tc/core"
+	"github.com/josharian/native"
+	"github.com/vishvananda/netns"
+	"golang.org/x/sys/unix"
+
+	"istio.io/pkg/log"
+)
+
+const (
+	FilesystemTypeBPFFS = unix.BPF_FS_MAGIC
+	MapsRoot            = "/sys/fs/bpf"
+	MapsPinpath         = "/sys/fs/bpf/ambient"
+
+	QdiscKind            = "clsact"
+	TcaBpfFlagActDiretct = 1 << 0 // refer to include/uapi/linux/pkt_cls.h TCA_BPF_FLAG_ACT_DIRECT
+	TcPrioFilter         = 1      // refer to include/uapi/linux/pkt_sched.h TC_PRIO_FILLER
+)
+
+var isBigEndian = native.IsBigEndian
+
+type RedirectServer struct {
+	redirectArgsChan           chan *RedirectArgs
+	obj                        ambient_redirectObjects
+	ztunnelHostingressFd       int
+	ztunnelHostingressProgName string
+	ztunnelIngressFd           int
+	ztunnelIngressProgName     string
+	inboundFd                  int
+	inboundProgName            string
+	outboundFd                 int
+	outboundProgName           string
+}
+
+func (r *RedirectServer) initBpfObjects() {
+	var options ebpf.CollectionOptions
+	if _, err := os.Stat(MapsPinpath); err != nil {
+		if os.IsNotExist(err) {
+			if err := os.Mkdir(MapsPinpath, os.ModePerm); err != nil {
+				log.Fatalf("unable to create ambient bpf mount directory: %v", err)
```

Comment:
return an error; generally we never want to exit in code, only by returning error up to the top level